### DISCUSSION TITLE
[MIGHT EXIST ALREADY] CP2K PLUGINS

### DISCUSSION THREAD

**Discussion by carajillu** [cp2k/cp2k#3764] [Ideas]
Dear all,
I am developing custom functionality for CP2K and so far I've been modifying core files such as src/motion/input_cp2k_md.F and src/motion/md_run.F, so I was wondering if there is less invasive way to implement custom functionality. I am wondering if I can put my code in a src/mystuff/ folder in a way similar to the implementation of custom CVs and biasing methods in plumed (see https://github.com/carajillu/plumed2_carajillu/tree/master/src/GHOSTPROBE for example). I realise Plumed is a C++ code and making use of the different classes makes this easier, but perhaps we can do the same thing with modules?
Thanks

**Comment by fstein93** [comment]
I am not sure what you mean by "developing custom functionality". What functionalities of CP2K do you need? What do you intend to implement? Where do you intend to implement your functionality?

**Comment by carajillu** [comment]
I'm trying to add a new subsection called MOTION>MD>CUSTOM_FORCE_EVAL,
which every x MD steps (controlled by a STRIDE keyword) will calculate
energies and forces using a method different than the one on the main
FORCE_EVAL section and output the results in a specific format.

For context, the goal of it is to run classical MD and generate ab initio
energies and forces of the systems every x MD steps, which will then be
used to train ML force fields. I know I could do that in a more sequential
way with REFTRAJ, but the idea is to integrate it in a workflow where,
after running the MD simulation, my training data is ready and in a format
suitable for the ML tool I am going to use.

Best,
Joan

*Mr. Joan Clark-Nicolas*

**Comment by fstein93** [comment]
This sounds similar to what was done [here](https://www.cp2k.org/_media/exercises:2017_uzh_cp2k-tutorial:rybkin_cp2k-tutorial-zurich2017.pdf).
